doc/specs/#476 - Pinned packages.md

```diff
@@ -0,0 +1,186 @@
+---
+author: Easton Pillay jedieaston/easton@planeteaston.com
+created on: 2022-01-25
+last updated: 2022-02-02
+issue id: 476
+---
+
+# Pinning a package
+
+For [#476](https://github.com/microsoft/winget-cli/issues/476)
+
+## Abstract
+
+This spec describes the functionality behind "pinning" a package (freezing a package at a certain version so that it will be not be automatically upgraded via `winget upgrade --all`).
+
+## Inspiration
+
+This functionality is inspired by functionality in other package managers, as well as community feedback. Many packages (software development tools and libraries commonly, but other software as well) introduce breaking changes that users may not want integrate into their workflow quite yet.
+
+## Solution Design
+
+A table of packages (by Name, Version and Product Code if available) that are currently pinned will be kept in the local tracking catalog. To be repository independent, these values should be correlated from local sources (Add and Remove Programs, Appx, etc). This also means that the user should be able to pin packages that were not installed via the Windows Package Manager, so that they won't be upgraded via `winget upgrade --all`. 
+
+
+## UI/UX Design
+
+Several commands will need to be added to support pinning packages. Where applicable, equivalent names (or new arguments) for PowerShell cmdlets will also be given.
+
+The main interface for the pinning feature will be a new command, `winget pin`, with a couple of options:
+
+
+```
+winget pin
+PowerShell: Get-WinGetPinnedPackage
```

Comment:
@denelon Should this be a separate cmdlet or a argument for `Get-WinGetPackage`? I couldn't decide.

Reply:
I would be voting for `Get-WinGetPackage -Pinned`😁

Reply:
The PowerShell convention is more likely to be discrete cmdlets for pinning and unpinning. Looking at approved PowerShell verbs: https://docs.microsoft.com/en-us/powershell/scripting/developer/cmdlet/approved-verbs-for-windows-powershell-commands?view=powershell-7.2

We may also consider Lock and Unlock, but that might be confusing when compared with the "pin" name.

I've thrown a few suggestions below. I'll also mention @RDMacLachlan to get some input here:

Get-WinGetPinnedPackage - shows packages that have been pinned.
Set-WinGetPinnedPackage - pins a package
Clear-WinGetPinnedPackage - unpins a package

We may also want to consider packages that upgrade themselves as not pinnable.

Reply:
I had a few other suggestions below earlier; I like `Clear` more than `Remove`. I know you already mentioned lock/unlock, but I think different nouns could be used.

`Lock-WinGetPackageVersion` and `Unlock-WinGetPackageVersion` - From [Common verbs](https://docs.microsoft.com/en-us/dotnet/api/System.Management.Automation.VerbsCommon?view=powershellsdk-1.1.0)
`Disable-WinGetPackageUpgrade` and `Enable-WinGetPackageUpgrade` - From [Lifecycle verbs](https://docs.microsoft.com/en-us/dotnet/api/System.Management.Automation.VerbsLifeCycle?view=powershellsdk-1.1.0)
`Suspend-WingetPackageUpgrade` and `Resume-WinGetPackageUpgrade` - From [Lifecycle verbs](https://docs.microsoft.com/en-us/dotnet/api/System.Management.Automation.VerbsLifeCycle?view=powershellsdk-1.1.0)

Note: After more consideration, I do like Set and Clear the best, since it maintains consistency